src/app/Library/CrudPanel/Traits/Fields.php

```diff
@@ -58,6 +58,8 @@ public function makeSureFieldHasNecessaryAttributes($field)
         $field = $this->makeSureFieldHasType($field);
         $field = $this->makeSureSubfieldsHaveNecessaryAttributes($field);
 
+        $this->setupFieldValidation($field, $field['parentFieldName'] ?? false);
```

Comment:
Won't this basically do `setValidation()` (aka run the validation) A LOT of times? Every time a field has validation? This is the bit that makes the final `setValidation()` call unnecessary, correct? That's its only purpose?

Reply:
This will call the `setValidation()` multiple times **BUT DO NOT RUN VALIDATION MULTIPLE TIMES**. 

The `setValidation` method is a switch to add different kinds of validation in the panel:

```php
public function setValidation($classOrRulesArray = false, $messages = [])
    {
        if (! $classOrRulesArray) {
            $this->setValidationFromFields();
        } elseif (is_array($classOrRulesArray)) {
            $this->setValidationFromArray($classOrRulesArray, $messages);
        } elseif (is_string($classOrRulesArray) || is_class($classOrRulesArray)) {
            $this->setValidationFromRequest($classOrRulesArray);
        } else {
            abort(500, 'Please pass setValidation() nothing, a rules array or a FormRequest class.');
        }
    }
```

So basically we call it in every field/subfield with an `array of rules`, so we will go through `setValidationFromArray` that will be aggregatting all field validations inside operation setting `validationRules`. **Not running it**.

The validate part is only called once in `Update/Create` operation, when we create the anonymous class, merge the form request with the field rules, and validate everything.

Reply:
Hmm... ok... I understand now, thank. I guess it's inefficient to do this upon every `addField()`... but then again... if that addField has a `required()`... you'd expect right after it you can query which fields are required and get it... soo... ok... 😀

Reply:
Ok so while testing this I realised something. This DOES eliminate the need to run `CRUD::setValidation()` at the end, for field-attribute-validation. BUUUT. **It only does that when you're using the array syntax.** For the fluent syntax... it does not.

So in `BadgeCrudController` for example, this will work just fine:
```php
        CRUD::addField([
            'name' => 'name',
            'validationRules' => 'required',
        ]);
```

But this will NOT work:
```php        
        CRUD::field('name')->validationRules('required');
```

It's clear why - because when the field is added using the fluent syntax, the validation rule isn't there. It's only added afterwards. 

This is a big inconsistency between the array and fluent syntax - I can totally see how developers will mistakenly use the fluent syntax and think it'll validate, then be surprised when it's not (or not even notice). So we gotta fix this.

We can easily make this work using the fluent syntax too, by adding the `validationRules()` and `validationMessages()` methods to the CrudField, which on top of setting those attributes on the field would also run `setupFieldValidation()`. Which would also fix #4284 I just introduced a few minutes ago 😅

Let's create a separate PR with this line and the above, pointing to this branch. It's a separate feature after all - it eliminates the need for `CRUD::setValidation()` after array and field-attribute validation. That'll make it easier to test and merge both this PR and that one.